FIAT/hdiv_trace.py

```diff
@@ -119,71 +173,109 @@ def tabulate(self, order, points, entity=None):
-        of derivative evaluations, are not mathematically well-defined.
+           Performing illegal tabulations on this element will result in either
+           a tabulation table of `numpy.nan` arrays (`entity=None` case), or
+           insertions of the `TraceError` exception class. This is due to the
+           fact that performing cell-wise tabulations, or asking for any order
+           of derivative evaluations, are not mathematically well-defined.
         """
-        facet_dim = self.ref_el.get_spatial_dimension() - 1
-        sdim = self.space_dimension()
-        nf = self.facet_element.space_dimension()
+        sd = self.ref_el.get_spatial_dimension()
+        facet_sd = sd - 1
 
         # Initializing dictionary with zeros
         phivals = {}
         for i in range(order + 1):
-            alphas = mis(self.ref_el.get_spatial_dimension(), i)
+            alphas = mis(sd, i)
+
             for alpha in alphas:
-                phivals[alpha] = np.zeros(shape=(sdim, len(points)))
-        evalkey = (0,) * (facet_dim + 1)
+                phivals[alpha] = np.zeros(shape=(self.space_dimension(),
+                                                 len(points)))
+
+        evalkey = (0,) * sd
 
         # If entity is None, identify facet using numerical tolerance and
         # return the tabulated values
         if entity is None:
+            # NOTE: Numerical approximation of the facet id is currently only
+            # implemented for simplex reference cells.
+            if not self.ref_el.get_shape() in [TRIANGLE, TETRAHEDRON]:
+                raise NotImplementedError(
+                    "Tabulating this element on a %s cell without providing "
+                    "an entity is not currently supported." % type(self.ref_el)
+                )
+
             # Attempt to identify which facet (if any) the given points are on
             vertices = self.ref_el.vertices
             coordinates = barycentric_coordinates(points, vertices)
-            (unique_facet, success) = extract_unique_facet(coordinates)
+            unique_facet, success = extract_unique_facet(coordinates)
 
             # If successful, insert evaluations
             if success:
                 # Map points to the reference facet
-                new_points = map_to_reference_facet(points, vertices, unique_facet)
+                new_points = map_to_reference_facet(points,
+                                                    vertices,
+                                                    unique_facet)
+
+                # Retrieve values by tabulating the DG element
+                element = self.dg_elements[facet_sd]
+                nf = element.space_dimension()
+                nonzerovals, = itervalues(element.tabulate(order, new_points))
+
+                sindex = nf * unique_facet
+                eindex = nf * (unique_facet + 1)
+                phivals[evalkey][sindex:eindex, :] = nonzerovals
```

Comment:
For the `entity=None` case, this bit doesn't work for tensor product cells. A few lines above, I make sure the reference cell is triangle/tetrahedron.